processing/src/main/java/io/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -0,0 +1,391 @@
+      );
+    }
+
+    /**
+     * No synchronization, because {@link #specialize} is called only from {@link #classSpecializationExecutor}, i. e.
+     * from a single thread.
+     */
+    byte[] getPrototypeClassBytecode() throws IOException
+    {
+      if (prototypeClassBytecode == null) {
+        ClassLoader cl = prototypeClass.getClassLoader();
+        try (InputStream prototypeClassBytecodeStream =
+                 cl.getResourceAsStream(prototypeClassBytecodeName + ".class")) {
+          prototypeClassBytecode = ByteStreams.toByteArray(prototypeClassBytecodeStream);
+        }
+      }
+      return prototypeClassBytecode;
+    }
+
+    private static String classBytecodeName(String className)
+    {
+      return className.replace('.', '/');
+    }
+  }
+
+  private static class SpecializationId
+  {
+    private final String runtimeShape;
+    private final ImmutableMap<Class<?>, Class<?>> classRemapping;
+    private final int hashCode;
+
+    private SpecializationId(String runtimeShape, ImmutableMap<Class<?>, Class<?>> classRemapping)
+    {
+      this.runtimeShape = runtimeShape;
+      this.classRemapping = classRemapping;
+      this.hashCode = runtimeShape.hashCode() * 1000003 + classRemapping.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+      if (!(obj instanceof SpecializationId)) {
+        return false;
+      }
+      SpecializationId other = (SpecializationId) obj;
+      return runtimeShape.equals(other.runtimeShape) && classRemapping.equals(other.classRemapping);
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return hashCode;
+    }
+  }
+
+  static class WindowedLoopIterationCounter<T> extends SpecializationState<T> implements Runnable
+  {
+    private final PerPrototypeClassState<T> perPrototypeClassState;
+    private final SpecializationId specializationId;
+    private final ConcurrentMap<Long, AtomicLong> perMinuteIterations = new ConcurrentHashMap<>();
+    private final AtomicBoolean specializationScheduled = new AtomicBoolean(false);
+
+    WindowedLoopIterationCounter(
+        PerPrototypeClassState<T> perPrototypeClassState,
+        SpecializationId specializationId
+    )
+    {
+      this.perPrototypeClassState = perPrototypeClassState;
+      this.specializationId = specializationId;
+    }
+
+    @Nullable
+    @Override
+    public T getSpecialized()
+    {
+      return null;
+    }
+
+    @Override
+    public void accountLoopIterations(long loopIterations)
+    {
+      if (specializationScheduled.get()) {
+        return;
+      }
+      if (loopIterations > triggerSpecializationIterationsThreshold ||
+          addAndGetTotalIterations(loopIterations) > triggerSpecializationIterationsThreshold) {
+        if (specializationScheduled.compareAndSet(false, true)) {
+          classSpecializationExecutor.submit(this);
+        }
+      }
+    }
+
+    private long addAndGetTotalIterations(long newIterations)
```

Comment:
can u add a comment, that it does so for last one hour... took me a bit to figure out